src/bluez/adapter.rs

```diff
@@ -0,0 +1,125 @@
+use super::peripheral::Peripheral;
+use crate::api::{BDAddr, Central, CentralEvent};
+use crate::{Error, Result};
+use async_trait::async_trait;
+use bluez_async::{AdapterId, BluetoothError, BluetoothEvent, BluetoothSession, DeviceEvent};
+use futures::stream::{self, Stream, StreamExt};
+use std::pin::Pin;
+
+#[derive(Clone, Debug)]
+pub struct Adapter {
+    session: BluetoothSession,
+    adapter: AdapterId,
+}
+
+impl Adapter {
+    pub(crate) fn new(session: BluetoothSession, adapter: AdapterId) -> Self {
+        Self { session, adapter }
+    }
+}
+
+#[async_trait]
+impl Central for Adapter {
+    type Peripheral = Peripheral;
+
+    async fn events(&self) -> Result<Pin<Box<dyn Stream<Item = CentralEvent>>>> {
+        // There's a race between getting this event stream and getting the current set of devices.
+        // Get the stream first, on the basis that it's better to have a duplicate DeviceDiscovered
+        // event than to miss one. It's unlikely to happen in any case.
+        let events = self.session.event_stream().await?;
+
+        // Synthesise `DeviceDiscovered' events for existing peripherals.
+        let devices = self.session.get_devices().await?;
+        let initial_events = stream::iter(
+            devices
+                .into_iter()
+                .map(|device| CentralEvent::DeviceDiscovered(BDAddr::from(&device.mac_address))),
+        );
```

Comment:
I'm a little confused by this. We emit device discovered *every* time we get an event stream from the adapter?

Reply:
Would using an `btleplug::api::AdapterManager`  to track which devices are new, and which need to be updated instead? Much like how the [`winrt`](https://github.com/qwandor/btleplug/blob/688b9f2e661d6f72182ca04e0609e029e0f59952/src/winrtble/adapter.rs#L49-L60) and [`corebluetooth`](https://github.com/qwandor/btleplug/blob/688b9f2e661d6f72182ca04e0609e029e0f59952/src/corebluetooth/adapter.rs#L47-L69) (Although, admittidly, corebluetooth doesn't appear to emit `DeviceUpdated` events).

Reply:
Currently yes. That's why I was asking on Discord exactly what behaviour people expect for devices which the system already knows about before `events()` is called. I still prefer leaving it up to the user to call `peripherals()` if they want to get existing devices, but this approach is more consistent with what the other platforms do. I assume calling `events()` multiple times is not how we expect people to use the API in most cases anyway.